### DISCUSSION TITLE
How is `htmlId` dangerous?

### DISCUSSION THREAD

**Discussion by nschloe** [KaTeX/KaTeX#4001] [Q&A]
In https://katex.org/docs/supported.html#html, one reads

> The following "raw HTML" features are potentially dangerous for untrusted inputs, [...]

I'm especially interested in `\htmlId`. Could you explain how this is could be dangerous? I could perhaps think of

- ID collisions with other elements
- Generic IDs like `"nav"` might interact with CSS/JS in unintended ways

Those problems would be mitigted by forcing a prefix/prefixes on the ID, or forcing it to comply to a given regex. Does this functionality exist in KaTeX yet?

**Comment by edemaine** [comment]
Yes, I had in mind ID collisions with other elements, which would cause `#id` links to change behavior on a page, `document.querySelector`/`document.getElementById` to behave differently in JS, and might trigger undesired CSS styling. I guess especially with the JS effects, this could be dangerous: if the JS code on the page assumes that some form input has a particular ID and that it's already validated, making a new object with the same ID could break these implicit assumptions. At the least, it's pretty easy to cause pages to break.

KaTeX does not have a feature for prefixing IDs or classes. It's an interesting idea. I could see a few approaches:

1. Add options for prefixing all ids and classes
2. Allow `trust` to return a special value like `{ prefix: 'katex-' }`. This would enable some ids to stay the same, and others to be prefixed, and allow the prefixes to be dynamic.
3. Allow `trust` to more generally modify the id/class/style, by returning a new string perhaps. This would let it sanitize `style` values instead of just saying "untrusted", allow adding suffixes or prefixes, etc.

I'm inclined to go with Option 3, as it's the most flexible, and avoids adding yet more options.

**Reply by delvh** [reply]
Option 3 sounds good to me.
On a related note: What's the danger behind `\htmlStyle`? After all, it only sets the style of the current element if I see that correctly.
Thus, isn't the "weird styling" the intended behavior?

**Reply by edemaine** [reply]
If you have an interest in Option 3, do you want to take a stab at implementing it?

With `style` you can make an element go all over the page, with e.g. `position: fixed`. This can cause a user post with KaTeX content to completely disrupt the entire page.
